pkg/services/oauthtoken/oauth_token.go

```diff
@@ -126,76 +135,92 @@ func (o *Service) HasOAuthEntry(ctx context.Context, usr identity.Requester) (*l
+func (o *Service) TryTokenRefresh(ctx context.Context, usr identity.Requester) (*oauth2.Token, error) {
 	ctxLogger := logger.FromContext(ctx)
 
 	if usr == nil || usr.IsNil() {
 		ctxLogger.Warn("Can only refresh OAuth tokens for existing users", "user", "nil")
 		// Not user, no token.
-		return nil
+		return nil, nil
 	}
 
 	if !usr.IsIdentityType(claims.TypeUser) {
 		ctxLogger.Warn("Can only refresh OAuth tokens for users", "id", usr.GetID())
-		return nil
+		return nil, nil
 	}
 
 	userID, err := usr.GetInternalID()
 	if err != nil {
 		ctxLogger.Warn("Failed to convert user id to int", "id", usr.GetID(), "error", err)
-		return nil
+		return nil, nil
 	}
 
 	ctxLogger = ctxLogger.New("userID", userID)
 
 	lockKey := fmt.Sprintf("oauth-refresh-token-%d", userID)
-	if _, ok := o.cache.Get(lockKey); ok {
-		ctxLogger.Debug("Expiration check has been cached, no need to refresh")
-		return nil
+
+	lockTimeConfig := serverlock.LockTimeConfig{
+		MaxInterval: 30 * time.Second,
+		MinWait:     160 * time.Millisecond, // p95 for token refresh is ~800ms, so at least it should wait 800ms before it fails (5 retries with (at least) 160 ms)
+		MaxWait:     250 * time.Millisecond,
 	}
-	_, err, _ = o.singleFlightGroup.Do(lockKey, func() (any, error) {
-		ctxLogger.Debug("Singleflight request for getting a new access token", "key", lockKey)
+
+	retryOpt := func(attempts int) error {
+		if attempts < 5 {
+			return nil
+		}
+		return ErrCouldntRefreshToken
+	}
+
+	var newToken *oauth2.Token
+	var cmdErr error
+
+	lockErr := o.serverLock.LockExecuteAndReleaseWithRetries(ctx, lockKey, lockTimeConfig, func(ctx context.Context) {
+		ctxLogger.Debug("Serverlock request for getting a new access token", "key", lockKey)
 
 		authInfo, exists, err := o.HasOAuthEntry(ctx, usr)
 		if !exists {
 			if err != nil {
 				ctxLogger.Debug("Failed to fetch oauth entry", "error", err)
-			} else {
-				// User is not logged in via OAuth no need to check
-				o.cache.Set(lockKey, struct{}{}, maxOAuthTokenCacheTTL)
 			}
-			return nil, nil
+			return
 		}
 
-		_, needRefresh, ttl := needTokenRefresh(authInfo)
+		storedToken, needRefresh := needTokenRefresh(authInfo)
 		if !needRefresh {
-			o.cache.Set(lockKey, struct{}{}, ttl)
-			return nil, nil
+			// Set the token which is returned by the outer function in case there's no need to refresh the token
+			newToken = storedToken
+			return
 		}
 
 		// get the token's auth provider (f.e. azuread)
 		provider := strings.TrimPrefix(authInfo.AuthModule, "oauth_")
 		currentOAuthInfo := o.SocialService.GetOAuthInfoProvider(provider)
 		if currentOAuthInfo == nil {
 			ctxLogger.Warn("OAuth provider not found", "provider", provider)
-			return nil, nil
+			return
 		}
 
 		// if refresh token handling is disabled for this provider, we can skip the refresh
 		if !currentOAuthInfo.UseRefreshToken {
 			ctxLogger.Debug("Skipping token refresh", "provider", provider)
-			return nil, nil
+			return
```

Comment:
I think we should be able to do this check before we lock. We have usr.GetAuthenticatedBy that should be able to tell of what provider was used to authenticate the user

Reply:
That way we don't have to aquire the lock when we are not authenticated with oauth or the provider has not enabled this feature

Reply:
Done